text/083-boost-stdlib-dev/text.md

```diff
@@ -0,0 +1,786 @@
+
+# Motivation
+
+Acknowledging that:
+
+- Many parts of the current standard library
+  are no longer *standard*, including parts whose use is discouraged.
+
+- The stdlib remains used in many, if not most of, Coq developments.
+
+- The development process of the stdlib seems in practice mostly stuck
+  for years: additions or changes are not openly encouraged and the
+  few wannabe contributors that are bold enough to open PRs often see
+  them left unattended.
+
+- The Coq repository may look intimidating for someone interested in
+  contributing to the library: heavy validation processes,
+  lots of OCaml code.
+
+- The teams interested in developing Coq itself and the stdlib are
+  largely disjoint, with most Coq developers having neither interest
+  nor time to significantly contribute to the stdlib.
+  
+We aim at:
+
+- Making sure that the stdlib components that people actively rely on
+  are efficiently maintained by interested parties.
+
+- Ensuring that potential new contributors to the library feel welcome.
+
+- Enabling the development of an evolved or different official "stdlib"
+  which embraces what could be agreed as current best practices for
+  library development.
+
+- Enabling the sharing of a larger basis between important general
+  libraries in the ecosystem.
+
+- Being sure that users are properly discouraged to rely on parts of the
+  stdlib that are considered subpar, especially when recommended
+  alternatives are available through the Coq Platform.
+
+To this end, we would like to
+ensure that the stdlib does not enjoy special status and that Coq can be
+used without it. We should consider stdlib components as libraries to
+advertise for their own values as part of the Coq Platform packages, but
+without their historical origin, or their development and release process
+being tied to Coq, (mis)leading users to consider them as the only
+officially recommended libraries.
+
+In particular, we should:
+
+- Identify consistent stdlib components that can be used independently
+  from each other and that would be worth distributing as separate
+  packages. Identify their maintainers and give them freedom to define
+  the future of the components they maintain, in the limits set by the
+  Coq Platform charter. Allow maintainers to extract stdlib components
+  to maintain and evolve them outside the core Coq repository and to have
+  their own release schedule and versioning scheme, in case they wish to
+  do so.
+
+- Extract the prelude + a minimum set of components that alternative
+  general libraries like MathComp and coq-stdpp need as a basis.
+  Make sure that this reduced core stdlib component can evolve to allow
+  other libraries to use newer features of Coq (like universe polymorphism,
+  SProp and primitive projections).
+  Stop including any other stdlib components as part of the `coq` opam
+  package and encourage maintainers of Coq packages in other distributions
+  to do the same.
```

Comment:
I think that the packaging story needs to be clarified. DEB-based distributions have a `libcoq-stdlib` package, while RPM-based distributions have a `coq` package, which contains the standard library (the actual binaries are in `coq-core`, which is a dependency). They also have packages that depend on those, e.g., for Flocq, `libcoq-flocq` and `flocq`, respectively. This CEP suggests to tell the maintainers of these distributions that, instead of maintaining a single package, they now have to maintain 50 different packages.

We should make sure that this extra maintenance work will not have an adverse effect on the general availability of Coq, in particular the disappearance of existing Coq packages from distributions. (This is currently one of the strong points of Coq over Lean.) For example, I know of a few universities (starting with mine), where software for teaching purpose needs only a few days to be installed if it is just one `apt` or `dnf` command away. But if there is no official package, then you better fill a formal request one month in advance to hopefully get software to be compiled and installed in time on all the computers used by students.

By the way, to a lesser extent, the packaging concern also applies to Opam, as the official repository is currently collapsing under its own weight. A working group has been put in motion to try to devise solutions to this issue, but most of the proposed solutions seem to revolve around the idea of removing a bunch of packages from the repository (or at least archiving them away from the repository). Needless to say, the sudden appearance of 50 new Opam packages is presumably not what they had in mind.

Even regarding Coq's Platform, some clarifications are needed. It currently proposes three levels of installation, "base", "full", and "extended", as well as some manual selection of packages, and Coq's standard library is part of all of these levels. This CEP begs a few questions:
- Would the standard library be removed from "base"? Or would it stay there, perhaps only partially?
- What about "full"? Would it contain the equivalent of the current standard library or only the parts that are needed by the packages installed by "full"?
- What about the manual selection of packages? Would the user have to select among the 50 packages of the current standard library or would some larger categories be defined to hide this complexity?

Reply:
Note that the packages are not really "different" packages, since they all depend on other packages of the exact same version, they can almost be trated as a single package. It's only a way to offer users to be explicit about the fact that they use only part of the global package. It's opt-in for users, who can keep using the metapackage with everything.
You can see @SnarkBoojum answer above about Debian

> Would the standard library be removed from "base"? Or would it stay there, perhaps only partially?

The library is currently part of "base", the current CEP doesn't offer to change that.

> What about "full"? Would it contain the equivalent of the current standard library or only the parts that are needed by the packages installed by "full"?

I guess, as the name implies "full" should include "base", again this CEP doesn't offer any change here.

> What about the manual selection of packages? Would the user have to select among the 50 packages of the current standard library or would some larger categories be defined to hide this complexity?

There is a metapackage. Not sure if something more fine grained would be needed. Ccing @msoegtrop and @rtetley in case they have a more precise opinion.

Reply:
I'm the maintainer for the Coq ecosystem in Debian and the main problems I encounter are:
- upstreams insisting they won't release a new version before a privileged distribution asks them for one (Coq Platform, generally);
- upstreams not being quite live: not officially abandoned projects, but the maintainer have so much on their plates they don't really have the time to keep them current.

Because even if the packages come from different projects, they are still globally tied: a new package for Coq breaking some other seven steps away means the Coq package is considered unsuitable. Case in point, I still couldn't upload Coq 8.19.1 to Debian, because the current status of the whole is:
```
* Update Coq [4/7]
 - [X] step 1 [3/3]
   - [X] elpi 1.18.2-1
   - [X] coq 8.19.1+dfsg-1
   - [X] coq-doc 8.19.1-1
 - [-] step 2 [15/17]
   - [X] aac-tactics 8.19.0-1
   - [X] coq-bignums 9.0.0+coq8.19-1
   - [X] coq-dpdgraph 1.0+8.19-1 ?
   - [X] coq-elpi 2.1.0-1
   - [X] coq-ext-lib 0.12.1-1
   - [X] coq-hammer 1.3.2+8.19-1
   - [X] coq-hott 8.19-1
   - [X] coq-libhyps 2.0.8-2
   - [X] coq-menhirlib 20231231+ds-2
   - [X] coq-record-update 0.3.4-1
   - [X] coq-reduction-effects 0.1.5-2
   - [X] coq-stdpp 1.10.0-1
   - [ ] coq-unicoq FAILURE [[https://github.com/unicoq/unicoq/issues/90][upstream report]]
   - [X] coq-unimath 20240331-1
   - [X] flocq 4.1.4-1
   - [ ] ott FAILURE [[https://github.com/ott-lang/ott/issues/105][upstream report]]
   - [X] paramcoq 1.1.3+coq8.19-1
 - [-] step 3 [6/8]
   - [X] coq-equations 1.3-8.19-1
   - [X] coq-gappa 1.5.5-1
   - [X] coq-hierarchy-builder 1.7.0-1
   - [X] coq-iris 4.2.0-1
   - [ ] coq-math-classes FAILURE [[https://github.com/coq-community/math-classes/issues/124][upstream report]]
   - [ ] coq-mtac2 WAIT coq-unicoq
   - [X] coq-simple-io 1.9.0-1
   - [X] coqprime 8.19-1
 - [-] step 4 [1/2]
   - [X] ssreflect 2.2.0-1
   - [ ] coq-corn WAIT coq-math-classes
 - [X] step 5 [8/8]
   - [X] coq-deriving 0.2.0-2
   - [X] coq-reglang 1.2.1-2
   - [X] coq-relation-algebra 1.7.10-2
   - [X] coquelicot 3.4.1-2
   - [X] mathcomp-bigenough 1.0.1-13
   - [X] mathcomp-finmap 2.1.0-1
   - [X] mathcomp-zify 1.5.0+2.0+8.16-2
   - [X] coq-quickchick 2.0.3-1
 - [X] step 6 [6/6]
   - [X] coq-extructures 0.4.0-2
   - [X] coq-interval 4.10.0-1
   - [X] mathcomp-algebra-tactics 1.2.3-2
   - [X] mathcomp-analysis 1.0.0-2
   - [X] mathcomp-multinomials 2.2.0-2
   - [X] mathcomp-real-closed 2.0.2-2
 - [X] step 7 [1/1]
   - [X] coqeal 2.0.2-1
```

Reply:
> Note that the packages are not really "different" packages, since they all depend on other packages of the exact same version, they can almost be treated as a single package.

This seems to contradict the previous goal of CEP, which explicitly says "allow maintainers to have their own release schedule and versioning scheme". As soon as various components of the standard library have differing release schedules, it becomes very difficult to treat it as a single package.